File: csharp/Behavioral/22-Visitor/Element/CarpentryJob.cs
namespace Visitor;

public class CarpentryJob : IJobElement
{
    public string JobId { get; } = Guid.NewGuid().ToString("N")[..8];

    public string Title { get; }
    public decimal Cost { get; }
    public decimal TimberMetres { get; }
    public string FinishType { get; }

    public CarpentryJob(string title, decimal cost, decimal timberMetres, string finishType)
    {
        Title        = title;
        Cost         = cost;
        TimberMetres = timberMetres;
        FinishType   = finishType;
    }

    public void Accept(IJobVisitor visitor) => visitor.Visit(this);
}

File: typescript/Behavioral/22-Visitor/Element/CarpentryJob.ts
import type { IJobElement } from "./IJobElement";
import type { IJobVisitor } from "../Visitor/IJobVisitor";

export class CarpentryJob implements IJobElement {
  readonly jobId: string = Math.random().toString(36).slice(2, 10);

  constructor(
    readonly title: string,
    readonly cost: number,
    readonly timberMetres: number,
    readonly finishType: string,
  ) {}

  accept(visitor: IJobVisitor): void {
    visitor.visitCarpentry(this);
  }
}
